app/vfe-vdpa/main.c

```diff
@@ -727,6 +728,113 @@ cmdline_parse_ctx_t main_ctx[] = {
 	NULL,
 };
 
+static int
+virtio_ha_client_start(void)
+{
+	struct virtio_dev_name *pf_list = NULL;
+	struct vdpa_vf_with_devargs *vf_list = NULL;
+	struct vdpa_vf_ctx *vf_ctx = NULL;
+	struct virtio_pf_ctx pf_ctx;
+	int ret, i, j, nr_pf, nr_vf;
+
+	ret = virtio_ha_ipc_client_init();
+	if (ret) {
+		RTE_LOG(ERR, VDPA, "Failed to init ha ipc client\n");
+		return -1;		
+	}
+
+	ret = virtio_ha_pf_list_query(&pf_list);
+	if (ret < 0) {
+		RTE_LOG(ERR, VDPA, "Failed to query pf list\n");
+		return -1;		
+	} else if (ret == 0) {
+		RTE_LOG(INFO, VDPA, "Query success: 0 pf to restore\n");
+		return 0;
+	} else {
+		RTE_LOG(INFO, VDPA, "Query success: %d pf to restore\n", ret);
+	}
+
+	nr_pf = ret;
+
+	for (i = 0; i < nr_pf; i++) {
+		ret = virtio_ha_pf_ctx_query(pf_list + i, &pf_ctx);
+		if (ret < 0) {
+			RTE_LOG(ERR, VDPA, "Failed to query pf ctx\n");
+			return -1;
+		}
+
+		ret = virtio_ha_pf_ctx_set(pf_list + i, &pf_ctx);
+		if (ret < 0) {
+			RTE_LOG(ERR, VDPA, "Failed to set pf ctx in pf driver\n");
+			return -1;
+		}
+
+		ret = rte_vdpa_pf_dev_add(pf_list[i].dev_bdf);
+		if (ret < 0) {
+			RTE_LOG(ERR, VDPA, "Failed to restore pf\n");
+			return -1;
+		}
+
+		ret = virtio_ha_pf_ctx_unset(pf_list + i);
+		if (ret < 0) {
+			RTE_LOG(ERR, VDPA, "Failed to unset pf ctx in pf driver\n");
+			return -1;
+		}
+	}
+
+	for (i = 0; i < nr_pf; i++) {
+		ret = virtio_ha_vf_list_query(pf_list + i, &vf_list);
+		if (ret < 0) {
+			RTE_LOG(ERR, VDPA, "Failed to query vf list of %s\n", pf_list[i].dev_bdf);
+			return -1;		
+		} else {
+			RTE_LOG(INFO, VDPA, "Query success: %d vf of pf %s to restore\n",
+				ret, pf_list[i].dev_bdf);
+		}
+
+		nr_vf = ret;
+
+		for (j = 0; j < nr_vf; j++) {
+			ret = virtio_ha_vf_ctx_query(&vf_list[j].vf_name, pf_list + i, &vf_ctx);
+			if (ret < 0) {
+				RTE_LOG(ERR, VDPA, "Failed to query vf ctx\n");
+				return -1;
+			}
+
+			ret = virtio_ha_vf_ctx_set(&vf_list[j].vf_name, vf_ctx);
+			if (ret < 0) {
+				RTE_LOG(ERR, VDPA, "Failed to set vf ctx in vf driver\n");
+				return -1;
```

Comment:
should free vf_ctx and vf_list if failed

Reply:
done